notary-commons/src/main/kotlin/sidechain/iroha/consumer/IrohaConsumerImpl.kt

```diff
@@ -72,50 +81,68 @@ class IrohaConsumerImpl(
      * @return list of hex hashes that were accepted by iroha
      */
     override fun send(lst: Iterable<TransactionOuterClass.Transaction>): Result<List<String>, Exception> {
-        irohaAPI.transactionListSync(lst)
         return Result.of {
-            lst.map {
-                Utils.hash(it)
-            }
-        }.map { hashes ->
-            hashes.map {
-                checkTransactionStatus(it)
+            val successfulTxHashesList = ArrayList<String>()
+            irohaAPI.transactionListSync(lst)
+            lst.map { tx -> Utils.hash(tx) }.forEach { txHash ->
+                val txStatus = TxStatus.createEmpty()
+                val observer = createTxStatusObserver(txStatus)
+                irohaAPI.txStatus(txHash).blockingSubscribe(observer)
+                if (txStatus.txException != null) {
+                    logger.error("Iroha batch error", txStatus.txException!!)
+                } else if (txStatus.txHash != null) {
+                    successfulTxHashesList.add(txStatus.txHash!!)
+                }
             }
-                .filter { res ->
-                    res.fold(
-                        { true },
-                        {
-                            logger.warn("Batch tx was failed: ", it)
-                            false
-                        }
-                    )
-                }.map { it.get() }
+            successfulTxHashesList
         }
     }
 
     /**
-     * Check if transaction is committed to Iroha
-     * @param hash - hash of transaction to check
-     * @return hex representation of hash or failure
+     * Create tx status observer
+     * @param txStatus - object that will hold tx status after observer completion
+     * @return tx status observer
      */
-    private fun checkTransactionStatus(hash: ByteArray): Result<String, Exception> {
-        return Result.of {
-            val responseObservable = irohaAPI.cmdStub.statusStream(Utils.createTxStatusRequest(hash))
-            val hex = String.hex(hash)
-            responseObservable.forEachRemaining { statusResponse ->
-                if (statusResponse.txStatus == Endpoint.TxStatus.STATEFUL_VALIDATION_FAILED) {
-                    val message =
-                        "Iroha transaction $hex received STATEFUL_VALIDATION_FAILED ${statusResponse.errOrCmdName}"
-                    logger.error { message }
-                    throw Exception(message)
-                }
+    private fun createTxStatusObserver(txStatus: TxStatus): InlineTransactionStatusObserver? {
```

Comment:
is there a way to return `Result` ?

Reply:
Yeah. But I don't think we need it there.